lib/connections/svc.go

```diff
@@ -19,71 +19,89 @@ import (
 	"github.com/syncthing/syncthing/lib/config"
 	"github.com/syncthing/syncthing/lib/discover"
 	"github.com/syncthing/syncthing/lib/events"
-	"github.com/syncthing/syncthing/lib/model"
 	"github.com/syncthing/syncthing/lib/protocol"
-	"github.com/syncthing/syncthing/lib/relay"
-	"github.com/syncthing/syncthing/lib/relay/client"
 
 	"github.com/thejerf/suture"
 )
 
-type DialerFactory func(*url.URL, *tls.Config) (*tls.Conn, error)
-type ListenerFactory func(*url.URL, *tls.Config, chan<- model.IntermediateConnection)
+type intermediateConnection struct {
+	*tls.Conn
+	Type     string
+	Priority int
+}
+
+type Connection struct {
+	intermediateConnection
+	protocol.Connection
+}
+
+type genericDialer interface {
+	Dial(protocol.DeviceID, *url.URL, *tls.Config) (intermediateConnection, error)
+	Priority() int
+}
+
+type listenerFactory func(*url.URL, *tls.Config, chan intermediateConnection) genericListener
+
+type genericListener interface {
+	Serve()
+	Stop()
+	// Addresses which can potentially be mutated for peers to be able to connect.
+	// Also, a single listen address might yield multiple addresses due to finding
+	// multiple gateways, etc.
+	WANAddresses() []*url.URL
+	LANAddresses() []*url.URL
+	Error() error
+}
 
 var (
-	dialers   = make(map[string]DialerFactory, 0)
-	listeners = make(map[string]ListenerFactory, 0)
+	dialers   = make(map[string]genericDialer, 0)
+	listeners = make(map[string]listenerFactory, 0)
 )
 
 type Model interface {
 	protocol.Model
-	AddConnection(conn model.Connection)
+	AddConnection(conn Connection)
 	ConnectedTo(remoteID protocol.DeviceID) bool
 	IsPaused(remoteID protocol.DeviceID) bool
 }
 
-// The connection service listens on TLS and dials configured unconnected
+// Svc is the connection service listens on TLS and dials configured unconnected
 // devices. Successful connections are handed to the model.
-type connectionSvc struct {
+type Svc struct {
```

Comment:
You scolded me for introducing the `connectionSvc` (or some other ...svc), and I can kind of swallow those, but just `Svc` is unnecessarily terse. Lets call it Service. I've been sort of meaning to refactor s/Svc/Service/ on everything else as well...